From fa58ebbd52d606eb862da7b6337b27dcf99f186e Mon Sep 17 00:00:00 2001
From: Rajan Vaja <rajan.vaja@xilinx.com>
Date: Tue, 7 Apr 2020 02:13:49 -0700
Subject: [PATCH 1323/1852] clk: fixed-factor: Don't register fixed factor clk
 until parent clk is registered

commit 514ae8d9d52d5b6695c90806df061ebd59d3620b from
https://github.com/Xilinx/linux-xlnx.git

"ecbf3f1795fd (clk: fixed-factor: Let clk framework find parent)" commit
registers fixed factor clock even if parent clock is not registered.

So when consumer tries to use fixed factor clock, it's parent may not
have registered, which may create issue while getting parent rate and
that will result in 0 clock rate for that fixed factor clock.

To fix this, do not register fixed factor clock until parent clock is
registered.

Signed-off-by: Rajan Vaja <rajan.vaja@xilinx.com>
State: not-upstreamable
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/clk/clk-fixed-factor.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/clk-fixed-factor.c b/drivers/clk/clk-fixed-factor.c
index 8b343e59dc61..3d14334591a4 100644
--- a/drivers/clk/clk-fixed-factor.c
+++ b/drivers/clk/clk-fixed-factor.c
@@ -163,6 +163,7 @@ static struct clk_hw *_of_fixed_factor_clk_setup(struct device_node *node)
 {
 	struct clk_hw *hw;
 	const char *clk_name = node->name;
+	const char *parent_name = NULL;
 	unsigned long flags = 0;
 	u32 div, mult;
 	int ret;
@@ -180,6 +181,9 @@ static struct clk_hw *_of_fixed_factor_clk_setup(struct device_node *node)
 	}
 
 	of_property_read_string(node, "clock-output-names", &clk_name);
+	parent_name = of_clk_get_parent_name(node, 0);
+	if (!parent_name)
+		return ERR_PTR(-EPROBE_DEFER);
 
 	if (of_match_node(set_rate_parent_matches, node))
 		flags |= CLK_SET_RATE_PARENT;
@@ -211,8 +215,8 @@ void __init of_fixed_factor_clk_setup(struct device_node *node)
 {
 	_of_fixed_factor_clk_setup(node);
 }
-CLK_OF_DECLARE(fixed_factor_clk, "fixed-factor-clock",
-		of_fixed_factor_clk_setup);
+CLK_OF_DECLARE_DRIVER(fixed_factor_clk, "fixed-factor-clock",
+		      of_fixed_factor_clk_setup);
 
 static int of_fixed_factor_clk_remove(struct platform_device *pdev)
 {
-- 
2.31.1

